### PR TITLE
[sonic-cfggen]: Update UT to add port lanes

### LINKED ISSUE #9553: [yang-models] port validation issue, lanes is mandatory
'PORT': {
		'Ethernet0': {
			'speed': '40000', 
			'description': 'OCPSCH0104012MS:Ethernet12', 
			'mtu': '9100', 
			'tpid': '0x8100', 
			'pfc_asym': 'off', 
			'admin_status': 'up'
		}
	}, 
Config db schema generated by minigraph can’t pass yang validation, PORT table does not have 'lanes' field.
It seems that we need to update yang model or sonic-config-engine.

#### Steps to reproduce the issue:
1. Run command 'sonic-cfggen -m tests/sample_graph.xml --print-data'.
2. Use sonic_yang to validate config db schema.

#### Describe the results you received:
Yang validation failed.

#### Describe the results you expected:
Yang validation should pass.

#### Output of `show version`:

```
(paste your output here)
```
#### Output of `show techsupport`:

```
(paste your output here or download and attach the file here )
```

### LINKED ISSUE #9579: [yang-models] port validation issue, speed is mandatory
#### Description

 'PORT': {'Ethernet0': {'alias': 'fortyGigE0/0',
                        'description': 'fortyGigE0/0',
                        'lanes': '29,30,31,32',
                        'mtu': '9100',
                        'pfc_asym': 'off',
                        'tpid': '0x8100'},
          'Ethernet100': {'admin_status': 'up',
                          'alias': 'fortyGigE0/100',
                          'description': 'Servers100:eth0',
                          'lanes': '125,126,127,128',
                          'mtu': '9100',
                          'pfc_asym': 'off',
                          'tpid': '0x8100'},

Config db schema generated by minigraph can’t pass yang validation:

* there's no 'speed' in port table, and it's mandatory.

It seems that we need to update sonic-config-engine or yang model.

#### Steps to reproduce the issue:
1. Run command 'sonic-cfggen -m tests/t0-sample-graph.xml -p tests/t0-sample-port-config.ini --print-data'.
2. Use sonic_yang to validate config db schema.

#### Describe the results you received:
Yang validation failed.

#### Describe the results you expected:
Yang validation should pass.

#### Output of `show version`:

```
(paste your output here)
```
#### Output of `show techsupport`:

```
(paste your output here or download and attach the file here )
```

### PR DESCRIPTION
**Pull request by ganglyu**
Signed-off-by: Gang Lv ganglv@microsoft.com

#### Why I did it
Need to run yang validation for sonic-cfggen unit test, and many unit test does not provide lanes for port table.

#### How I did it
Update port config file.

#### How to verify it
Run sonic-cfggen unit test,
Use below PR to verify 
https://github.com/Azure/sonic-buildimage/pull/10228/

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog

Fix #9553

### CODE REVIEW

File: src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml [hunk 1]
```diff
@@ -847,6 +847,54 @@
...
+          <PortName>0</PortName>
+          <Priority>0</Priority>
+          <Speed>40000</Speed>
+        </a:EthernetInterface>
+        <a:EthernetInterface>
+          <ElementType>DeviceInterface</ElementType>
+          <AlternateSpeeds i:nil="true"/>
+          <EnableFlowControl>true</EnableFlowControl>
+          <Index>1</Index>
+          <InterfaceName>Ethernet4</InterfaceName>
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml:891]: >Ethernet4 [](http://example.com/codeflow?start=25&length=9)

Why Ethernet1/../4 are needed? I checked device/arista/x86_64-arista_7050_qx32s/Arista-7050QX-32S/port_config.ini. And seems they are not used. And alias Ethernet4 will conflict with SONiC port name. I am not sure if it is problem.

**Comment by qiluo-msft** [comment]
>         <EndPort>Ethernet1</EndPort>

Do you need to change this line at the same time?

---
Refers to: src/sonic-config-engine/tests/sample_graph.xml:115 in f313124. [](commit_id = f31312412c4755706f9a27bb0b56d2f422f85ce0, deletion_comment = False)

File: src/sonic-config-engine/tests/test_j2files.py [hunk 2]
```diff
@@ -326,16 +326,18 @@ def test_swss_switch_render_template(self):
...
                 "graph": self.t0_minigraph,
+                "port": self.t0_port_config,
                 "output": "t0-switch.json"
             },
         }
         for _, v in test_list.items():
-            argument = " -m {} -y {} -t {} > {}".format(
-                v["graph"], constants_yml, switch_template, self.output_file
+            argument = " -m {} -p {} -y {} -t {} > {}".format(
+                v["graph"], v["port"], constants_yml, switch_template, self.output_file
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_j2files.py:340]: >port [](http://example.com/codeflow?start=31&length=4)

port_config is more readable.

File: src/sonic-config-engine/tests/test_multinpu_cfggen.py [hunk 3]
```diff
@@ -80,68 +81,73 @@ def test_dummy_run(self):
...
+            output = self.run_script_for_asic(argument, asic, self.port_config[asic])
             self.assertEqual(output.strip(), 'value1\nvalue2')
 
     def test_metadata_tacacs(self):
-        argument = '-m "' + self.sample_graph +  '" --var-json "TACPLUS_SERVER"'
+        argument = '-m "' + self.sample_graph + '" -p "' + self.sample_port_config +  '" --var-json "TACPLUS_SERVER"'
         output = json.loads(self.run_script(argument))
         self.assertDictEqual(output, {'123.46.98.21': {'priority': '1', 'tcp_port': '49'}})
         #TACPLUS_SERVER not present in the asic configuration.
+        argument = '-m "' + self.sample_graph +  '" --var-json "TACPLUS_SERVER"'
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_multinpu_cfggen.py:130]: >argument [](http://example.com/codeflow?start=8&length=8)

If you do not add this line, will it fail?

File: src/sonic-config-engine/tests/test_multinpu_cfggen.py [hunk 4]
```diff
@@ -297,23 +303,38 @@ def test_device_asic_metadata(self):
...
-                             'SNMP_ACL':  {'policy_desc': 'SNMP_ACL',    'services': ['SNMP'],        'stage': 'ingress', 'type': 'CTRLPLANE'},
-                             'SSH_ONLY':  {'policy_desc': 'SSH_ONLY',    'services': ['SSH'],         'stage': 'ingress', 'type': 'CTRLPLANE'}})
+        exp = {\
+                'SNMP_ACL': {'policy_desc': 'SNMP_ACL', 'type': 'CTRLPLANE', 'stage': 'ingress', 'services': ['SNMP']},
+                'EVERFLOW': {'policy_desc': 'EVERFLOW', 'stage': 'ingress', 'ports': ['PortChannel0002', 'PortChannel0008', 'Ethernet8', 'Ethernet12', 'Ethernet24', 'Ethernet28'], 'type': 'MIRROR'},
+                'EVERFLOWV6': {'policy_desc': 'EVERFLOWV6', 'stage': 'ingress', 'ports': ['PortChannel0002', 'PortChannel0008', 'Ethernet8', 'Ethernet12', 'Ethernet24', 'Ethernet28'], 'type': 'MIRRORV6'},
+                'SSH_ONLY': {'policy_desc': 'SSH_ONLY', 'type': 'CTRLPLANE', 'stage': 'ingress', 'services': ['SSH']},
+                'DATAACL': {'policy_desc': 'DATAACL', 'stage': 'ingress', 'ports': ['PortChannel0002', 'PortChannel0008'], 'type': 'L3'}}
+        for k, v in output.items():
+            if 'ports' in v and len(v['ports']):
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_multinpu_cfggen.py:315]: >len [](http://example.com/codeflow?start=32&length=3)

No need to check len.

File: src/sonic-config-engine/tests/multi_npu_data/sample_port_config.ini [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# name          lanes            speed    alias           index    asic_port_name    role 
+Ethernet0       33,34,35,36      10000    Ethernet1/1        0       Eth0-ASIC0         Ext
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/multi_npu_data/sample_port_config.ini:2]: >   [](http://example.com/codeflow?start=67&length=2)

Align the columns?

File: src/sonic-config-engine/tests/sample_output/py2/l2switch.json [hunk 6]
```diff
@@ -4,162 +4,194 @@
...
-            "admin_status": "up"
+            "admin_status": "up",
+            "speed": "10000"
         },
         "Ethernet4": {
             "alias": "fortyGigE0/4",
             "lanes": "25,26,27,28",
-            "admin_status": "up"
+            "admin_status": "up",
+            "speed": "10000"
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/sample_output/py2/l2switch.json:20]: >speed [](http://example.com/codeflow?start=13&length=5)

Top question: in concept, there should be some default port speed. You may check swss implementation. Maybe we need to fix src/sonic-yang-models/yang-models/sonic-port.yang?

Comment (qiluo-msft) [reply src/sonic-config-engine/tests/sample_output/py2/l2switch.json:20]: Missing port speed in config_db is by design. ref: [https://github.com/Azure/SONiC/wiki/L2-Switch-mode](https://github.com/Azure/SONiC/wiki/L2-Switch-mode)

File: src/sonic-config-engine/tests/sample_output/py2/ports.json [hunk 7]
```diff
@@ -20,18 +20,207 @@
         },
         "OP": "SET"
     },
+    {
+        "PORT_TABLE:Ethernet108": {
+            "speed": "10000",
+            "description": "fortyGigE0/108"
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/sample_output/py2/ports.json:26]: What will fail if you do not add?

File: src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml (hunk 1 above)
Comment (ganglyu) [reply src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml:891]: There's no speed for these ports. 
We shouldn't update device/arista/x86_64-arista_7050_qx32s/Arista-7050QX-32S/port_config.ini, so I update this xml.

**Comment by ganglyu** [comment]
> > ```
> >     <EndPort>Ethernet1</EndPort>
> > ```
> 
> Do you need to change this line at the same time?
> 
> Refers to: src/sonic-config-engine/tests/sample_graph.xml:115 in [f313124](https://github.com/Azure/sonic-buildimage/commit/f31312412c4755706f9a27bb0b56d2f422f85ce0). [](commit_id = [f313124](https://github.com/Azure/sonic-buildimage/commit/f31312412c4755706f9a27bb0b56d2f422f85ce0), deletion_comment = False)

There's no Ethernet1 in port_config.ini, so I replace it with Ethernet4

File: src/sonic-config-engine/tests/test_multinpu_cfggen.py (hunk 3 above)
Comment (ganglyu) [reply src/sonic-config-engine/tests/test_multinpu_cfggen.py:130]: I add port config in line 126, and below line will also add port config:
output = json.loads(self.run_script_for_asic(argument, asic, self.port_config[asic]))
If don't add line 130, there will be two port config for the same command.

File: src/sonic-config-engine/tests/sample_output/py2/l2switch.json (hunk 6 above)
Comment (ganglyu) [reply src/sonic-config-engine/tests/sample_output/py2/l2switch.json:20]: then port speed is not mandatory, I will update yang models instead.

File: src/sonic-config-engine/tests/sample_output/py2/ports.json (hunk 7 above)
Comment (ganglyu) [reply src/sonic-config-engine/tests/sample_output/py2/ports.json:26]: https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-config-engine/tests/test_j2files.py#L55-L59
This case will fail

File: src/sonic-config-engine/tests/test_cfggen.py [hunk 8]
```diff
@@ -394,38 +396,38 @@ def test_minigraph_port_autonegotiation(self):
...
-                "'Ethernet108': {'alias': 'fortyGigE0/108', 'pfc_asym': 'off', 'lanes': '81,82,83,84', 'description': 'fortyGigE0/108', 'mtu': '9100', 'tpid': '0x8100'}, "
-                "'Ethernet112': {'lanes': '89,90,91,92', 'description': 'ARISTA01T1:Ethernet1/1', 'pfc_asym': 'off', 'mtu': '9100', 'tpid': '0x8100', 'alias': 'fortyGigE0/112', 'admin_status': 'up'}, "
-                "'Ethernet116': {'lanes': '93,94,95,96', 'description': 'ARISTA02T1:Ethernet1/1', 'pfc_asym': 'off', 'mtu': '9100', 'tpid': '0x8100', 'alias': 'fortyGigE0/116', 'admin_status': 'up'}, "
-                "'Ethernet120': {'lanes': '97,98,99,100', 'description': 'ARISTA03T1:Ethernet1/1', 'pfc_asym': 'off', 'mtu': '9100', 'tpid': '0x8100', 'alias': 'fortyGigE0/120', 'admin_status': 'up'}, "
-                "'Ethernet124': {'lanes': '101,102,103,104', 'fec': 'rs', 'pfc_asym': 'off', 'mtu': '9100', 'tpid': '0x8100', 'alias': 'fortyGigE0/124', 'admin_status': 'up', 'speed': '100000', 'description': 'ARISTA04T1:Ethernet1/1'}}"
+                "{'Ethernet0': {'lanes': '29,30,31,32', 'alias': 'fortyGigE0/0', 'description': 'fortyGigE0/0', 'mtu': '9100', 'tpid': '0x8100', 'pfc_asym': 'off'}, "
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_cfggen.py:399]: >Ethernet0 [](http://example.com/codeflow?start=19&length=9)

Seems you don't need to change this hunk, you only change the order of fields. But they are converted to dict any way.

File: src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml (hunk 1 above)
Comment (qiluo-msft) [reply src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml]: If port speed is not mandatory, no need to fix?

File: src/sonic-config-engine/tests/test_cfggen.py (hunk 8 above)
Comment (ganglyu) [reply src/sonic-config-engine/tests/test_cfggen.py:399]: will try

File: src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml (hunk 1 above)
Comment (ganglyu) [reply src/sonic-config-engine/tests/sample-arista-7050-t0-minigraph.xml]: I have removed this change

File: src/sonic-config-engine/tests/sample_output/py2/l2switch.json (hunk 6 above)
Comment (qiluo-msft) [reply src/sonic-config-engine/tests/sample_output/py2/l2switch.json]: Sorry maybe I am wrong. I do see port speed in l2switch config. Need to double check.

File: src/sonic-config-engine/tests/test_cfggen.py [hunk 9]
```diff
@@ -563,44 +565,45 @@ def test_minigraph_neighbor_interfaces(self):
 
     def test_minigraph_neighbor_interfaces_config_db(self):
         # test to check if PORT table is retrieved from config_db
-        argument = '-m "' + self.sample_graph_simple_case + '" -v "PORT"'
+        argument = '-m "' + self.sample_graph_simple_case + '" -p "' + self.port_config + '" -v "PORT"'
         output = self.run_script(argument)
 
+        #self.maxDiff = None
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_cfggen.py:571]: >#self.maxDiff = None [](http://example.com/codeflow?start=8&length=20)

Remove dead code.

File: src/sonic-config-engine/tests/test_cfggen.py [hunk 10]
```diff
@@ -877,6 +879,7 @@ def test_minigraph_voq_inband_port(self):
         self.assertDictEqual(
             output_dict['Ethernet-IB0'],
             {'lanes': '133', 
+
             'alias': 'Recycle0',
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_cfggen.py:883]: Could you remove the unnecessary new line?

File: src/sonic-config-engine/tests/test_multinpu_cfggen.py [hunk 11]
```diff
@@ -215,17 +221,17 @@ def test_frontend_asic_ports(self):
...
-        argument = "-m {} -n asic0 --var-json \"PORT\"".format(self.sample_graph)
+        argument = "-m {} -p {} -n asic0 --var-json \"PORT\"".format(self.sample_graph, self.port_config[0])
         output = json.loads(self.run_script(argument))
         self.assertDictEqual(output,
-                {"Ethernet0": { "admin_status": "up",  "alias": "Ethernet1/1",  "asic_port_name": "Eth0-ASIC0",  "description": "01T2:Ethernet1:config_db",  "index": "0",  "lanes": "33,34,35,36",   "mtu": "9100", "tpid": "0x8100",   "pfc_asym": "off",  "role": "Ext",  "speed": "40000", "autoneg": "on" },
-                "Ethernet4": { "admin_status": "up",  "alias": "Ethernet1/2",  "asic_port_name": "Eth1-ASIC0",  "description": "01T2:Ethernet2:config_db",  "index": "1",  "lanes": "29,30,31,32",   "mtu": "9100", "tpid": "0x8100",  "pfc_asym": "off",  "role": "Ext",  "speed": "40000", "autoneg": "on" },
-                "Ethernet8": { "admin_status": "up", "alias": "Ethernet1/3",  "asic_port_name": "Eth2-ASIC0",  "description": "Ethernet1/3:config_db",  "index": "2",  "lanes": "41,42,43,44",   "mtu": "9100", "tpid": "0x8100",  "pfc_asym": "off",  "role": "Ext",  "speed": "40000" },
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_multinpu_cfggen.py:223]: Why admin_status disappear? Is it expected?

Comment (ganglyu) [reply src/sonic-config-engine/tests/test_multinpu_cfggen.py:223]: The actual output does not have admin_status

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**